proposals/stdlib/instant.md

```diff
@@ -0,0 +1,1365 @@
+
+### Migration process
+
+#### Constraints
+
+The API entries we have to take into account during the migration:
+
+* The `Instant` class itself.
+* The `Clock` interface.
+* `kotlinx-datetime` functions that accept a `Clock` as a parameter:
+  `Clock.todayIn(TimeZone)`, `Clock.asTimeSource()`.
+* `kotlinx-datetime` functions that return a `Clock`:
+  only `val Clock.System`.
+* `kotlinx-datetime` functions that accept an `Instant` as a parameter:
+  quite many of them.
+* `kotlinx-datetime` functions that return an `Instant`: also many.
+* `java.io.Serializable` implementation for `Instant`.
+* Third-party functions that accept an `Instant` as a parameter.
+* Third-party functions that return an `Instant`.
+* Places where `Instant` is set as a type parameter.
+  The only first-party example is serializers:
+  `InstantIso8601Serializer`, `InstantComponentSerializer`.
+* The serial descriptor names in serializers.
+
+There are two additional types of API entries that could require special care
+but luckily do not:
+
+* Overridable methods that accept an `Instant`.
+* Overridable methods that return an `Instant`.
+
+The only `open class` that `kotlinx-datetime` has is `TimeZone`, but
+its constructor is `internal`, so inheriting from it is not supported.
+
+The only `interface` that `kotlinx-datetime` has that mentions `Instant`
+is `Clock`.
+
+#### Unavoidable issues
+
+Regardless of the migration path taken,
+`kotlinx-datetime` will have to suppress errors to access `MIN` and `MAX`,
+which it needs to implement `Instant` parsing.
+As discussed in the "Behavior on overflow" and "Admissible ranges" sections,
+exposing them just for our internal needs is too high a price.
+However, this is not a hard requirement:
+`MIN` and `MAX` are only used for comparing to them, and we can check this via
+indirect and less efficient means:
+
+```kotlin
+fun epochSecondsLaterThanMaxInstant(epochSeconds: Long): Boolean =
+   Instant.fromEpochSeconds(epochSeconds) ==
+   Instant.fromEpochSeconds(epochSeconds - 1)
+```
+
+#### Process 1: pure library solutions
+
+* Standard Library publishes its own `kotlin.time.Instant`.
+* `kotlinx-datetime`:
+  - And adds conversion functions between `kotlinx.datetime.Instant`
+    and `kotlin.time.Instant`, as well as between `kotlinx.datetime.Clock` and
+    `kotlin.time.Clock`. These are public.
+  - Deprecates `kotlinx.datetime.Instant` and `kotlinx.datetime.Clock`
+    with a warning and a replacement with
+    `kotlin.time.Instant` and `kotlin.time.Clock`.
+  - Deprecates every function that accepts an `Instant`/`Clock`,
+    adds an overload that accepts `kotlin.time.(Instant|Clock)`
+    instead. The proposed suggestions involve the new converters.
+    Exception: `Clock.asTimeSource` does not get the new overload.
+  - For every function that returns but doesn't accept an `Instant`,
+    hides it and adds a new function, one that returns `kotlin.time.Instant`,
+    but with a different `JvmName`.
```

Comment:
I think a different `JvmName` isn't needed. Iirc two methods with the same name but different return types can exist in the JVM bytecode.

Reply:
Yes, you are right. For other platforms, though, it's still needed. Edited.